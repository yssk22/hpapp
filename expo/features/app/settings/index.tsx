import { logLogin, logLogout } from '@hpapp/system/firebase';
import * as logging from '@hpapp/system/logging';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Platform } from 'react-native';

import SettingsAppConfig, { AppConfig, SettingsAppConfigDefault } from './internals/SettingsAppConfig';
import SettingsCurrentUser, { CurrentUser } from './internals/SettingsCurrentUser';
import SettingsList from './internals/SettingsList';
import Provider, { useSettings } from './internals/SettingsProvider';
import SettingsUPFCConfig, { UPFCConfig } from './internals/SettingsUPFCConfig';
import SettingsUserConfig, { UserConfig, SettingsUserConfigDefault } from './internals/SettingsUserConfig';

export type SettingsProviderProps = {
  appConfig?: AppConfig;
  userConfig?: UserConfig;
  currentUser?: CurrentUser;
  upfcConfig?: UPFCConfig;
};

/**
 * A provider to setup the settings context
 */
export function SettingsProvider({
  appConfig,
  userConfig,
  currentUser,
  upfcConfig,
  children
}: SettingsProviderProps & {
  children: React.ReactNode;
}) {
  const [ready, setReady] = useState(false);
  useEffect(() => {
    (async () => {
      if (Platform.OS === 'web') {
        // we expect web is used only for storybook so we set the user token here from the environment variable and enforce the logged in user
        const accessToken = process.env['HPAPP_USER_TOKEN_FOR_JEST'];
        if (accessToken === undefined) {
          throw new Error('HPAPP_USER_TOKEN_FOR_JEST is not set for web');
        }
        await SettingsCurrentUser.save({
          id: '1234',
          username: 'yssk22',
          accessToken
        });
      } else {
        currentUser && (await SettingsCurrentUser.save(currentUser));
        userConfig && (await SettingsUserConfig.save(userConfig));
        appConfig && (await SettingsAppConfig.save(appConfig));
        upfcConfig && (await SettingsUPFCConfig.save(upfcConfig));
      }
      setReady(true);
    })();
  }, []);
  if (!ready) {
    return null;
  }
  return <Provider settings={SettingsList}>{children}</Provider>;
}

/**
 * Get the application configuration
 */
export function useAppConfig() {
  const [appConfig] = useSettings(SettingsAppConfig);
  return useMemo(() => {
    return {
      ...SettingsAppConfigDefault,
      ...appConfig
    };
  }, [appConfig]);
}

export { SettingsAppConfigDefault };

/**
 * Get the updator for AppConfig
 */
export function useAppConfigUpdator() {
  const [, updator] = useSettings(SettingsAppConfig);
  return updator;
}

/**
 * Get the user configuration
 */
export function useUserConfig() {
  const [userConfig] = useSettings(SettingsUserConfig);
  return userConfig;
}

/**
 * Get the updator for user configuration
 */
export function useUserConfigUpdator() {
  const [, updator] = useSettings(SettingsUserConfig);
  return updator;
}

/**
 * Get the UPFC configuration
 */
export function useUPFCConfig() {
  const [upfcConfig] = useSettings(SettingsUPFCConfig);
  // migrate from legacy helloproject credentials to new one to support mline
  return useMemo(() => {
    if (upfcConfig) {
      return {
        ...upfcConfig,
        username: undefined,
        password: undefined,
        // eslint-disable-next-line deprecation/deprecation
        hpUsername: upfcConfig?.username ?? upfcConfig?.hpUsername,
        // eslint-disable-next-line deprecation/deprecation
        hpPassword: upfcConfig?.password ?? upfcConfig?.hpPassword
      };
    }
    return undefined;
  }, [upfcConfig]);
}

/**
 * Get the updator for UPFC configuration
 */
export function useUPFCConfigUpdator() {
  const [, updator] = useSettings(SettingsUPFCConfig);
  return updator;
}

/**
 * Get the current user
 */
export function useCurrentUser() {
  const [currentUser] = useSettings(SettingsCurrentUser);
  return currentUser;
}

export type User = NonNullable<ReturnType<typeof useCurrentUser>>;

/**
 * Get the updator for current user
 */
export function useCurrentUserUpdator() {
  const [currentUser, updator] = useSettings(SettingsCurrentUser);
  return useCallback(
    (update: CurrentUser | null) => {
      if (currentUser) {
        if (update == null) {
          logging.Info('features.auth.logout', 'logout', {
            userid: currentUser.id
          });
          logLogout();
        }
      } else {
        if (update) {
          logging.Info('features.auth.login', 'login', {
            userid: update.id
          });
          // TODO: we should have a way to know the login method
          logLogin('default', update.id);
        }
      }
      return updator(update);
    },
    [updator]
  );
}

export { SettingsUserConfigDefault };
