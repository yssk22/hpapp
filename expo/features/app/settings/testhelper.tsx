import { act, render } from '@testing-library/react-native';

import { SettingsProvider, SettingsProviderProps } from './';
import SettingsList from './internals/SettingsList';

/**
 * Clear all settings
 */
export async function clearSettings() {
  await Promise.all(
    SettingsList.map((settings) => {
      settings.clear();
    })
  );
}

export async function renderSettingsProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  await clearSettings();
  const content = render(<SettingsProvider {...props}>{children}</SettingsProvider>);
  await act(async () => {});
  return content;
}
