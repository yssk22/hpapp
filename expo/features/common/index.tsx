import AmebloIcon from './internals/AmebloIcon';
import AssetIcon from './internals/AssetIcon';
import CalendarDateIcon from './internals/CalendarDateIcon';
import ConsentGate from './internals/ConsentGate';
import ErrorBoundary, { FallbackComponent } from './internals/ErrorBoundary';
import ExternalImage from './internals/ExternalImage';
import Initializer from './internals/Initializer';
import Link from './internals/Link';
import Loading from './internals/Loading';
import Text from './internals/Text';
import useAssetContent from './internals/useAssetContent';
import useAsync from './internals/useAsync';
import useErrorMessage from './internals/useErrorMessage';
import useIsMounted from './internals/useIsMounted';
import useMutationPromise from './internals/useMutationPromise';
import useReloadableAsync, { ReloadableAysncResult } from './internals/useReloadableAsync';
import useURICache from './internals/useURICache';

export {
  AmebloIcon,
  AssetIcon,
  CalendarDateIcon,
  ConsentGate,
  ErrorBoundary,
  ExternalImage,
  FallbackComponent,
  Initializer,
  Link,
  Loading,
  Text,
  useErrorMessage,
  useMutationPromise,
  useReloadableAsync,
  ReloadableAysncResult,
  useAssetContent,
  useAsync,
  useIsMounted,
  useURICache
};
