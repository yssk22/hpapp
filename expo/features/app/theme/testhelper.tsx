import { SettingsProviderProps } from '@hpapp/features/app/settings';
import { renderSettingsProvider } from '@hpapp/features/app/settings/testhelper';

import { ThemeProvider } from './';

export async function renderThemeProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  return renderSettingsProvider(<ThemeProvider>{children}</ThemeProvider>, props);
}
