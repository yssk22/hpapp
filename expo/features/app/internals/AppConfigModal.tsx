import { useAppConfig, useAppConfigUpdator } from '@hpapp/features/app/settings';
// import { Text } from '@hpapp/features/common';
import { FontSize, Spacing } from '@hpapp/features/common/constants';
import { Button, Dialog, Input, Switch } from '@rneui/themed';
import { useState } from 'react';
import { StyleSheet, View, Dimensions, Text } from 'react-native';

const deviceHeight = Dimensions.get('window').height;

type AppConfigModalProps = { isVisible: boolean; onClose: () => void };

/**
 * AppConfigModal shows a modal to configure the application settings.
 *
 * @param {boolean} visible - Indicates whether the modal is visible.
 * @param {() => void} onClose - Callback function to close the modal.
 */
export default function AppConfigModal({ isVisible, onClose }: AppConfigModalProps) {
  const appConfig = useAppConfig();
  const appConfigUpdate = useAppConfigUpdator();
  const [useStorybook, setUseStorybook] = useState(appConfig.useStorybook ?? false);
  const [useCustomGraphQLEndpoint, setUseCustomGraphQLEndpoint] = useState(appConfig.useCustomGraphQLEndpoint ?? false);
  const [graphQLEndpoint, setGraphQLEndpoint] = useState(appConfig.graphQLEndpoint ?? '');
  const [useLocalAuth, setUseLocalAuth] = useState(appConfig.useLocalAuth ?? false);
  const [useUPFCDemoScraper, setUseUPFCDemoScraper] = useState(appConfig.useUPFCDemoScraper ?? false);
  const onSave = () => {
    appConfigUpdate({
      useStorybook,
      useCustomGraphQLEndpoint,
      graphQLEndpoint,
      useLocalAuth,
      useUPFCDemoScraper
    });
    onClose();
  };
  return (
    <Dialog isVisible={isVisible} overlayStyle={styles.dialog}>
      <Dialog.Title>App Configuration</Dialog.Title>
      <View style={styles.content} testID="AppConfigModal.content">
        <View style={styles.form}>
          <View style={[styles.inputGroup, styles.switchContainer]}>
            <Text style={styles.label}>Use Storybook</Text>
            <Switch
              value={useStorybook}
              onValueChange={() => {
                setUseStorybook(!useStorybook);
              }}
            />
          </View>
          <View style={[styles.inputGroup, styles.switchContainer]}>
            <Text style={styles.label}>Use Local Auth</Text>
            <Switch
              value={useLocalAuth}
              onValueChange={() => {
                setUseLocalAuth(!useLocalAuth);
              }}
            />
          </View>
          <View style={[styles.inputGroup, styles.switchContainer]}>
            <Text style={styles.label}>Use Custom GraphQL Endpoint</Text>
            <Switch
              value={useCustomGraphQLEndpoint}
              onValueChange={() => {
                setUseCustomGraphQLEndpoint(!useCustomGraphQLEndpoint);
              }}
            />
          </View>
          <View style={!styles.inputGroup}>
            <Text style={styles.label}>GraphQL Endpoint</Text>
            <Input
              keyboardType="url"
              containerStyle={styles.input}
              errorStyle={styles.inputError}
              placeholder="http://localhost:8080/graphql/v3"
              value={graphQLEndpoint}
              onChangeText={(text) => {
                setGraphQLEndpoint(text);
              }}
              testID="AppConfigModal.graphQLEndpoint"
            />
          </View>
          <View style={[styles.inputGroup, styles.switchContainer]}>
            <Text style={styles.label}>Use UPFC Demo Scraper</Text>
            <Switch
              value={useUPFCDemoScraper}
              onValueChange={() => {
                setUseUPFCDemoScraper(!useUPFCDemoScraper);
              }}
            />
          </View>
        </View>
        <View style={styles.buttonGroup}>
          <Button type="outline" style={styles.button} onPress={onClose} testID="AppConfigModal.btnClose">
            Cancel
          </Button>
          <Button type="solid" style={styles.button} onPress={onSave} testID="AppConfigModal.btnSave">
            Save
          </Button>
        </View>
      </View>
    </Dialog>
  );
}

const styles = StyleSheet.create({
  dialog: {
    height: deviceHeight - deviceHeight * 0.15,
    width: '90%'
  },
  content: {
    flex: 1,
    flexDirection: 'column'
  },
  form: {
    flex: 1,
    flexGrow: 1
  },
  label: {
    paddingLeft: Spacing.Small,
    paddingRight: Spacing.Small
  },
  default: {
    fontStyle: 'italic',
    fontSize: FontSize.XSmall
  },
  inputGroup: {
    marginTop: Spacing.XSmall
  },
  switchContainer: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: Spacing.Small
  },
  input: {},
  inputError: {},
  footer: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  buttonGroup: {
    marginTop: Spacing.Small,
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  },
  button: {
    flexGrow: 1,
    paddingLeft: Spacing.Small,
    paddingRight: Spacing.Small
  }
});
