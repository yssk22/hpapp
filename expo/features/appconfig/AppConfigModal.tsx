import { useAppConfigForEdit } from '@hpapp/features/appconfig/useAppConfig';
import Text from '@hpapp/features/common/components/Text';
import { FontSize, Spacing } from '@hpapp/features/common/constants';
import { Button, Dialog, Input, Switch } from '@rneui/themed';
import { useCallback, useState } from 'react';
import { StyleSheet, View, Dimensions } from 'react-native';

const deviceHeight = Dimensions.get('window').height;

type AppConfigModalProps = { isVisible: boolean; onClose: () => void };

/**
 * AppConfigModal shows a modal to configure the application settings.
 *
 * @param {boolean} visible - Indicates whether the modal is visible.
 * @param {() => void} onClose - Callback function to close the modal.
 */
export default function AppConfigMoal({ isVisible, onClose }: AppConfigModalProps) {
  const [appConfig, setAppConfig] = useAppConfigForEdit();
  const [useLocalAppConfig, setUseLocalAppConfig] = useState(appConfig?.useLocalAppConfig ?? false);
  const [graphQLEndpoint, setGraphQLEndpoint] = useState(appConfig?.graphQLEndpoint ?? '');
  const [useLocalAuth, setUseLocalAuth] = useState(appConfig?.useLocalAuth ?? false);
  const onSave = useCallback(() => {
    setAppConfig({
      useLocalAppConfig,
      graphQLEndpoint,
      useLocalAuth
    });
    onClose();
  }, [appConfig, setAppConfig]);
  return (
    <Dialog isVisible={isVisible} overlayStyle={styles.dialog}>
      <Dialog.Title>App Configuration</Dialog.Title>
      <View style={styles.content}>
        <View style={styles.form}>
          <View style={[styles.inputGroup, styles.switchContainer]}>
            <Text style={styles.label}>Use Local AppConfig</Text>
            <Switch
              value={useLocalAppConfig}
              onValueChange={() => {
                setUseLocalAppConfig(!useLocalAppConfig);
              }}
            />
          </View>
          <View style={[styles.inputGroup, styles.switchContainer]}>
            <Text style={styles.label}>Use Local Auth</Text>
            <Switch
              value={useLocalAuth}
              onValueChange={() => {
                setUseLocalAuth(!useLocalAuth);
              }}
            />
          </View>
          <View style={!styles.inputGroup}>
            <Text style={styles.label}>GraphQL Endpoint</Text>
            <Input
              disabled={!useLocalAppConfig}
              keyboardType="url"
              containerStyle={styles.input}
              errorStyle={styles.inputError}
              placeholder="http://localhost:8080/graphql/v3"
              value={graphQLEndpoint}
              onChangeText={(text) => {
                setGraphQLEndpoint(text);
              }}
            />
          </View>
        </View>
        <View style={styles.buttonGroup}>
          <Button type="outline" style={styles.button} onPress={onClose}>
            Cancel
          </Button>
          <Button type="solid" style={styles.button} onPress={onSave}>
            Save
          </Button>
        </View>
      </View>
    </Dialog>
  );
}

const styles = StyleSheet.create({
  dialog: {
    height: deviceHeight - deviceHeight * 0.15,
    width: '90%'
  },
  content: {
    flex: 1,
    flexDirection: 'column'
  },
  form: {
    flex: 1,
    flexGrow: 1
  },
  label: {
    paddingLeft: Spacing.Small,
    paddingRight: Spacing.Small
  },
  default: {
    fontStyle: 'italic',
    fontSize: FontSize.XSmall
  },
  inputGroup: {
    marginTop: Spacing.XSmall
  },
  switchContainer: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: Spacing.Small
  },
  input: {},
  inputError: {},
  footer: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  buttonGroup: {
    marginTop: Spacing.Small,
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  },
  button: {
    flexGrow: 1,
    paddingLeft: Spacing.Small,
    paddingRight: Spacing.Small
  }
});
