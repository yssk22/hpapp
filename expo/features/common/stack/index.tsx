import { useThemeColor } from '@hpapp/features/app/theme';
import { isEmpty } from '@hpapp/foundation/string';
import {
  DefaultTheme,
  NavigationContainer,
  NavigationContainerRef,
  NavigationState as NavigationStateOrig,
  useNavigation as useNavigationOrig,
  useRoute
} from '@react-navigation/native';
import {
  createNativeStackNavigator as createNavigator,
  NativeStackNavigationOptions,
  NativeStackNavigationProp
} from '@react-navigation/native-stack';
import React, { forwardRef, useCallback, useEffect, useMemo } from 'react';

export type ScreenParams = { [key: string]: unknown };
export type Screen<P extends ScreenParams> = {
  name: string;
  path: string;
  component: React.ElementType<P>;
  options?: NativeStackNavigationOptions;
};
export type ScreenList = Screen<any>[];

type ComponentPropType<T> = T extends React.ComponentType<infer P> | React.Component<infer P> ? P : never;

type Route<P extends ScreenParams> = {
  key: string;
  name: string;
  path: string;
  params: P;
};

function screenComponent<P extends ScreenParams>(
  screen: Screen<P>,
  rootComponent: React.ElementType<{ children: React.ReactElement }>
) {
  const RootComponent = rootComponent;
  return function Screen() {
    const route = useRoute<Route<P>>();
    const props = route.params;
    const Component = screen.component;
    return <RootComponent>{React.createElement(Component, props)}</RootComponent>;
  };
}

type StackProps<P extends ScreenParams> = {
  initialRouteName: string;
  screens: ScreenList;
  onStateChange?: (args: { params: Readonly<object | undefined>; screen: Screen<P> }) => void;
} & Omit<ComponentPropType<ReturnType<typeof createNavigator>['Navigator']>, 'children'>;

function createStackNavigator<P extends ScreenParams>(
  props: {
    rootComponent?: React.ElementType<{ children: React.ReactElement }>;
  } = {}
) {
  const stack = createNavigator();
  const rootComponent = props.rootComponent ?? React.Fragment;
  return forwardRef(function Navigator(
    props: StackProps<P>,
    ref?: React.Ref<NavigationContainerRef<ReactNavigation.RootParamList>> | undefined
  ): JSX.Element {
    const [primary, contrastPrimary] = useThemeColor('primary');
    const [secondary] = useThemeColor('secondary');
    const { screens, onStateChange, initialRouteName, ...navigatorProps } = props;
    const screenMap = screens.reduce(
      (map, s) => {
        map.set(s.path, s);
        return map;
      },
      new Map() as Map<string, Screen<P>>
    );
    const handleStateChange = useCallback(
      (state: NavigationStateOrig | undefined) => {
        if (state) {
          const currentRoute = state.routes[state.index];
          onStateChange &&
            onStateChange({
              params: currentRoute.params ?? {},
              screen: screenMap.get(currentRoute.name)!
            });
        }
      },
      [onStateChange, screenMap]
    );
    const handleReady = useCallback(() => {
      const screen = screenMap.get(initialRouteName);
      if (screen) {
        onStateChange &&
          onStateChange({
            params: {},
            screen
          });
      }
    }, [initialRouteName, screenMap, onStateChange]);
    const Screens = useMemo(() => {
      return screens.map((screen) => {
        const options = screen.options ?? {};
        return (
          <stack.Screen
            key={screen.path}
            name={screen.path}
            component={screenComponent(screen, rootComponent)}
            options={{
              headerBackTitle: '',
              headerBackTitleVisible: false,
              headerTintColor: contrastPrimary,
              headerStyle: {
                backgroundColor: primary
              },
              ...options
            }}
          />
        );
      });
    }, [screens, primary, contrastPrimary]);
    return (
      <NavigationContainer
        ref={ref}
        onStateChange={handleStateChange}
        onReady={handleReady}
        theme={{
          ...DefaultTheme,
          colors: {
            primary,
            background: contrastPrimary,
            card: contrastPrimary,
            text: primary,
            border: contrastPrimary,
            notification: secondary
          }
        }}
      >
        <stack.Navigator initialRouteName={initialRouteName} {...navigatorProps}>
          {Screens}
        </stack.Navigator>
      </NavigationContainer>
    );
  });
}

const useNavigation = () => {
  const navigation = useNavigationOrig<NativeStackNavigationProp<Record<string, object | undefined>>>();
  const { navigate, replace, push, canGoBack, ...rest } = navigation;
  return useMemo(() => {
    function _navigate<P extends ScreenParams>(screen: Screen<P>, params?: P) {
      const path = screen.path;
      if (params) {
        navigate(path, params);
      } else {
        navigate(path);
      }
    }
    function _push<P extends ScreenParams>(screen: Screen<P>, params?: P) {
      const path = screen.path;
      if (params) {
        push(path, { initial: false, ...params });
      } else {
        push(path);
      }
    }

    function _replace<P extends ScreenParams>(screen: Screen<P>, params?: P) {
      const path = screen.path;
      if (params) {
        replace(path, params);
      } else {
        replace(path);
      }
    }

    function _canGoBack() {
      return canGoBack();
    }

    return {
      navigate: _navigate,
      push: _push,
      replace: _replace,
      canGoBack: _canGoBack,
      ...rest
    };
  }, [navigate]);
};

type Navigation = ReturnType<typeof useNavigation>;

function defineScreen<P extends ScreenParams>(path: string, component: React.ElementType<P>): Screen<P> {
  const name = (component as { name?: string }).name;
  if (isEmpty(name)) {
    // eslint-disable-next-line no-console
    console.warn(`Component (path: "${path}") should have a name, use path as a fallback`);
  }
  return {
    name: name ?? path,
    path,
    component,
    options: {}
  };
}

// shortcut for useNavigationOption({title: title})
function useScreenTitle(title: string) {
  const navigation = useNavigation();
  useEffect(() => {
    navigation.setOptions({
      title
    });
  }, [title]);
}

function useNavigationOption(options: Partial<NativeStackNavigationOptions>) {
  const navigation = useNavigation();
  useEffect(() => {
    navigation.setOptions(options);
  }, [options]);
}

export { createStackNavigator, useNavigation, Navigation, defineScreen, useScreenTitle, useNavigationOption };
