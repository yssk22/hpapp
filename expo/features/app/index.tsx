import { maybeDev } from '@hpapp/foundation/environment';
import { LocalMediaManagerProvider } from '@hpapp/system/media';

import AppRelayProvider from './internals/AppRelayProvider';
import { default as Root, AppRootProps as RootProps } from './internals/AppRoot';
import AppUpdateBanner from './internals/AppUpdateBanner';
import { ThemeProvider, ThemeProviderProps } from './theme';

export type AppRootProps = RootProps & ThemeProviderProps;

export function AppRoot({ screens, ...props }: AppRootProps) {
  return (
    <ThemeProvider {...props}>
      <AppRelayProvider>
        <LocalMediaManagerProvider name={maybeDev() ? 'hellofanapp-dev' : 'hellofanapp'}>
          <AppUpdateBanner />
          <Root screens={screens} />
        </LocalMediaManagerProvider>
      </AppRelayProvider>
    </ThemeProvider>
  );
}
