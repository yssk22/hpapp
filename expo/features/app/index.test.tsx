import { defineScreen } from '@hpapp/features/common/stack';
import { renderUserComponent, renderUserScreen, waitForUserRootRerendered } from '@hpapp/features/testhelper';
import { sleep } from '@hpapp/foundation/globals';
import { act, fireEvent, render, waitFor, waitForElementToBeRemoved } from '@testing-library/react-native';
import * as Updates from 'expo-updates';
import { Text } from 'react-native';

import { AppRoot } from './';
import AppConfigModal from './internals/AppConfigModal';
import { SettingsUserConfigDefault, useAppConfig } from './settings';
import { clearSettings } from './settings/testhelper';

const userConfig = {
  ...SettingsUserConfigDefault,
  completeOnboarding: true,
  consentOnToS: true,
  consentOnPrivacy: true,
  consentOnUPFCDataPolicy: true
};

beforeEach(async () => {
  await clearSettings();
});

describe('app', () => {
  const TestScreen = defineScreen('/', function TestScreen() {
    return <Text testID="app.test">Hello, Guest!</Text>;
  });

  describe('AppConfigModal', () => {
    test('render and close', async () => {
      const onClose = jest.fn();
      const content = await renderUserComponent(<AppConfigModal isVisible onClose={onClose} />);
      await act(() => {});
      expect(content.getByTestId('AppConfigModal.content')).toBeTruthy();
      await act(async () => {
        fireEvent.press(content.getByTestId('AppConfigModal.btnClose'));
      });
      expect(onClose).toHaveBeenCalled();
    });

    test('render and save', async () => {
      function AppConfigTestComponent() {
        const appConfig = useAppConfig();
        return <Text testID="app.test">{appConfig.graphQLEndpoint}</Text>;
      }

      const onClose = jest.fn();
      const content = await renderUserComponent(
        <>
          <AppConfigTestComponent />
          <AppConfigModal isVisible onClose={onClose} />
        </>
      );
      await act(() => {});
      const input = content.getByTestId('AppConfigModal.graphQLEndpoint');
      expect(input).toBeTruthy();
      await act(async () => {
        // fireEvent.changeText(input, 'https://risa.example.com:8080/graphql/v3'); // This line is not working
        input.props.onChangeText('https://risa.example.com:8080/graphql/v3');
      });
      await act(async () => {
        fireEvent.press(content.getByTestId('AppConfigModal.btnSave'));
      });
      // change is reflected in AppConfigSettings
      await waitForUserRootRerendered();
      expect(onClose).toHaveBeenCalled();
      expect(content.getByTestId('app.test').props.children).toEqual('https://risa.example.com:8080/graphql/v3');
    });
  });

  describe('AppUpdateBanner', () => {
    test('show banner', async () => {
      const fetchUpdateAsync = jest.fn();
      const reloadAsync = jest.fn();
      jest.spyOn(Updates, 'fetchUpdateAsync').mockImplementation(fetchUpdateAsync);
      jest.spyOn(Updates, 'reloadAsync').mockImplementation(reloadAsync);
      jest.spyOn(Updates, 'checkForUpdateAsync').mockImplementation(async () => {
        return {
          isAvailable: true,
          manifest: {
            id: 'test',
            commitTime: 100,
            assets: []
          },
          isRollBackToEmbedded: false,
          reason: undefined
        };
      });

      const content = await renderUserScreen(TestScreen);
      expect(content).toMatchSnapshot();
      const banner = await content.getByTestId('AppUpdateBanner');
      expect(banner).toBeTruthy();

      // hit the banner to trigger updates
      await act(async () => {
        fireEvent(banner, 'press');
      });
      expect(fetchUpdateAsync).toHaveBeenCalled();
      expect(reloadAsync).toHaveBeenCalled();

      // reloadAsync will reload the app so do not need to validate the end result of the banner

      // wait to surpress Animated(View) warning
      await waitFor(async () => {
        await sleep(500);
      });
    });
  });

  describe('AppRoot', () => {
    test('user', async () => {
      const content = await render(
        <AppRoot
          screens={[TestScreen]}
          currentUser={{
            id: '1',
            username: 'risa',
            accessToken: 'token'
          }}
          userConfig={userConfig}
        />
      );
      await act(() => {});
      expect(content.getByTestId('UserRoot.Loading')).toBeTruthy();
      await waitForElementToBeRemoved(() => content.queryByTestId('UserRoot.Loading'));
      expect(content.getByTestId('app.test')).toBeTruthy();
      expect(content.queryByTestId('AppRootGuest')).toBeNull();
    });

    test('guest', async () => {
      const content = await render(<AppRoot screens={[TestScreen]} userConfig={userConfig} />);
      await act(() => {});
      expect(content.queryByTestId('UserRoot.Loading')).toBeNull();
      expect(content.getByTestId('AppRootGuest')).toBeTruthy();
    });
  });
});
