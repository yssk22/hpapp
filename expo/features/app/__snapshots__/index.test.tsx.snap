// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app AppUpdateBanner show banner 1`] = `
[
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        fullScreenSwipeShadowEnabled={false}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        hasLargeHeader={false}
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onGestureCancel={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillAppear={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        sheetAllowedDetents="large"
        sheetCornerRadius={-1}
        sheetExpandsWhenScrolledToEdge={true}
        sheetGrabberVisible={false}
        sheetLargestUndimmedDetent="all"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            collapsable={false}
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "white",
                },
                undefined,
              ]
            }
          >
            <View
              style={
                {
                  "backgroundColor": "#ffffff",
                  "flex": 1,
                }
              }
            >
              <Text
                testID="app.test"
              >
                Hello, Guest!
              </Text>
            </View>
          </View>
        </View>
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backTitle=""
          backTitleVisible={false}
          backgroundColor="#0075c2"
          color="white"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="/"
          titleColor="white"
          topInsetEnabled={false}
          translucent={false}
        />
      </RNSScreen>
    </RNSScreenStack>
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#faad14",
        "flexDirection": "column",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 12,
        "paddingBottom": 8,
        "paddingTop": 8,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    testID="AppUpdateBanner"
  >
    <Text
      style={
        [
          {
            "color": "black",
          },
          {
            "fontSize": 14,
          },
          [
            {
              "fontSize": 14,
              "textAlign": "center",
            },
            {
              "color": "#ffffff",
            },
          ],
        ]
      }
    >
      アップデートが利用可能です。タップしてインストール。
    </Text>
  </View>,
]
`;
